### ISSUE TITLE
POS: Improve reconciliation loop in adverse network environments

### ISSUE THREAD

**Issue by blaggacao** [odoo/odoo#25666]
## Impacted versions:
all

## Steps to reproduce:
not reproducable, unlesss you can stage a specific DCHP environment, which would be:
(I can't so I'm suggesting)

- Setup your DHCP server so that it **_requires_** a dhcp lease release `dhclient -v -r eth0` from clients
- **Hard** Power Cycle the Raspberry Pi 

## Current behavior:

- The raspberry will not renew it's IP, as the (strictly configured) DHCP server denies new assignation without prior release of previous lease.

- `dhclient` can not read the persisted lease database on ~`/var/lib/dhcp/dhclient.leases`~ `/var/db/dhcpcd- interface -ssid .lease` (for `dhcpcd` as used on RPI) as `/var` is an ephemeral ramdisk

## Expected behavior:
In order to please a (strict) dhcp environment, read the lease database and reuse that lease if it's not expired so to avoid a new lease request, which would be negated by the server.

@pedrobaeza To avoid off-topic noise, please consider rather contacting me on Skype, if you need any explanation (skype: `blaggacao`). This report can be considered as a "as complete and precise as possible" description.

@Yenthe666 Who would be in charge of this? Do you know?

OPW: 1865188

**Comment by blaggacao** [comment]
## Analysis

Not sure if ...
https://github.com/odoo/odoo/blob/d6a71fae939e0f16302abaf24222692a275ee599/addons/point_of_sale/tools/posbox/overwrite_after_init/etc/dhcpcd.conf#L17
... also works on a really hard power cycle, or wether it has the desired effect. 

- I guess, in any case persistence files as described in the [dhcp docs](https://www.gsp.com/cgi-bin/man.cgi?section=8&topic=dhcpcd) would be needed. 

- Also `lastlease` might be a wise flag to add...

**Comment by Yenthe666** [comment]
Hmm, difficult one. As it is related to the point of sale I guess @pimodoo is a good start.

**Comment by pimodoo** [comment]
@jorenvo Could you advise us on this one?

**Comment by jorenvo** [comment]
I've never heard of DHCP servers behaving in this way, nor do I understand the benefit of forcing a client to release a valid lease. Why not just allow it to use it? What happens if a client somehow loses this lease because they clear some data? Then the client can no longer connect until that lease expires on the DHCP server? Could you point to some flag in isc-dhcp-server that enables this crazy policy?

**Comment by blaggacao** [comment]
> Why not just allow it to use it?

If my assessment is true, this is not the real problem. The real problem is: "How to make the network admin change that?" (assuming he is the type of self apprentice network admin usually present in SME)

> What happens if a client somehow loses this lease because they clear some data?

In my impure scenario, they would just say probably "change the cable and review the dust on the ventilators and don't forget to restart" :wink: 

> Then the client can no longer connect until that lease expires on the DHCP server?

Unfortunately I have limited (no!) support or insight from the network admin. After all, in this particular case, it might surface the limits of his knowledge (and I'm younger than him, it's an ego thing), but this is what my observation would permit to conclude.

> Could you point to some flag in isc-dhcp-server that enables this crazy policy?

What made me doubt was this paragraph:
>Tell dhclient to release the current lease it has from the server. This is not required by the DHCP protocol, but some ISPs require their clients to notify the server if they wish to release an assigned IP address.

I'm not sure what exact DHCP server they are using, nor what network equipment, but it fits the observation pretty well.

> nor do I understand the benefit of forcing a client to release a valid lease

My current hypothesis goes that the DHCP is denying a new lease (remember the valid lease is not persisted on `/var`) if it still holds registry of a previous valid lease. And only after some grace period, allows for a new lease to be offered. During this time, the Bos Box booted and already spanned a WIFI consequential in a general and hard to track* POS system failure (no POS/Restaurant invoicing).
In other words, the assumption is, that the DHCP is not graceful with hard power cycles.

* nobody might have gone that far, yet upon such observation. Pos BOX might just be qualified as "unstable".

**Comment by blaggacao** [comment]
**Add:** Implicitly the below flag allows for a DHCP not delivering a lease. Actually the observation allows for a different scenario: Let's say we have a power drop and the Pos Box recovers more quickly than the DHCP or the DHCP is in a remote location (eg. headquarter) and currently internet connection is faulty and they don't have local caches...

All those adverse scenarios could be mitigated by persisting the lease, I guess.

> -E -, -lastlease
> If dhcpcd cannot obtain a lease, then try to use the last lease acquired for the interface. If the -p, -persistent option is not given then the lease is used if it hasn’t expired.

**Comment by jorenvo** [comment]
The issue is that I don't think the behavior you describe exists. I don't think there's DHCP servers that prevent you from connecting until you release your old IP. If such behavior existed I think it would cause problems with all modern network clients, not just the POSBox.

Regardless, for the POSBox it's not possible to do because it can't store anything persistently. We do this on purpose to avoid:

1. corrupted filesystems when the POSBox is unplugged, it won't be cleanly powered off
2. write exhaustion on the SD card

The only exception we make is storing the Wifi configuration because it happens only rarely.

> Unfortunately I have limited (no!) support or insight from the network admin.

I understand that you are experiencing issues, but in depth debugging is necessary to solve it. Checking the logs of the dhcp client on the POSBox after it fails would be a good first step. An issue like this is tricky and in my opinion not possible to solve with guesswork.

> Add: Implicitly, the below flag allows for interpreting a scenario where a DHCP does not deliver a lease.

In this case the POSBox is not expected to work. A network managed with DHCP requires a working DHCP server. Even if you would enable these (dangerous) flags no other devices on the LAN would work, since no operating system falls back on such behavior by default.

> Let's say we have a power drop and the Pos Box recovers more quickly than the DHCP or

In case this happens indeed the POSBox might have to be restarted. In case you're using WIFI it would start AP mode with the POSBox ESSID. In case you're using a wired connection I think it's possible it might actually recover, but I would have to check if dhcpd eventually gives up or not.

> the DHCP is in a remote location (eg. headquarter) and currently internet connection is faulty and they don't have local caches while the hard power cycle occurs (because an unstable power supply eg. a inductive industrial consumer next door)...

If somehow the DHCP server becomes unreachable a POSBox is not expected to boot up and work. If it's already running it would remain working for a while until it's DHCP lease expires. This would once again affect all clients, not just the POSBox.

I don't think we'll be able to resolve this issue without more information.

**Comment by blaggacao** [comment]
>The issue is that I don't think the behavior you describe exists.

You might be right with that. Then again, a failure scenario which would mimic this hypothesis is easily constructed: Let's just assume it was a very short time failure in power supply coupled with a network recovery situation, where normal clients weren't affected because they where connected on a UPC, but the switch serving the PosBOX as well as the PosBox itself weren't.

This is a more robust scenario which does not depend on implementation details of a particular DHCP implementation. As I don't know the root cause (yet), this can fully replace the respective causalities in my previous reasoning.

>I don't think we'll be able to resolve this issue without more information.

I understand. Guesswork is not the best foundation. But still with a sharp minded discussion, we can reduce the trial and error phase. We also have to keep in mind the client impact and willingness (as I mentioned, it's not optimal).

> I would have to check if dhcpd eventually gives up or not.

This would be very useful as it gives us insight as to to when the point of no return is reached and therefore what is the maximum network outage that is supported gracefully. I somewhat tend to assume it's not more than a few seconds. At `rc.local`, AP creation already kicks in.


In unreliable IOT environments (I would argue: like this one), a unreachable network at the wrong time for a few seconds needs to be factored into the solution design. Without wanting to compromise your request or further investigation: Don't you think that should be a design principle?

To address your points (or give ideas how to work around them or find a better balance):
1. corrupted filesystems when the POSBox is unplugged, it won't be cleanly powered off
2. write exhaustion on the SD card

ad 1.: the probability of impact on a ext3/4 is already reduced 
ad 1.: the magnitude of impact can be reduced through a small (4mb mount, possibly disposable in case of error)
ad 1.: a reconciliation loop on startup can do `fdisk` on that mount and if it's definitely gone, dispose of it and create a new one.

ad 2:. This is a hard one... It would put an inconvenient TTL on the BOX itself

Maybe there is an **easier workaround** by just persisting a static IP and load that, just as for the AP. It would make the whole network behavior more predictable and *really* stateless (besides initial state).

I have no solution for transitioning back to AP and therefore manageability through a laptop. But I still feel, blocking this transition for the sake of stability is multiple times better than unexpected transitions at risk of a sudden a negative business impact.

Further reference: https://www.embeddedarm.com/about/resource/preventing-filesystem-corruption-in-embedded-linux

**Comment by jorenvo** [comment]
I understand where you're coming from but trying to make the POSBox work on a bad network I think is a battle we can't win. Keep in mind the case you mention would only happen when a POSBox is powered up and DHCP is unavailable. It can be resolved by simply rebooting the POSBox when DHCP is back. Maybe you can work around this very specific DHCP issue but what happens when someone scans a barcode and the LAN isn't working?

For your specific case coming up with a solution where the POSBox would recover on it's own is tricky. We don't want people to be completely locked out of their POSBox when they change a Wifi password or they move it to a different location. So for this we actually want it to start it's own AP when it fails to connect so people can connect and reconfigure, that's why it we did this in the first place.

**Comment by blaggacao** [comment]
Thanks @jorenvo for your insights! Actually, the problem can be re-framed:

As for BCM (business continuity management) ina POS application just one question is effectively relevant:

- Can a specific client X reach it's _corresponding_ RPi Y (MAC address)?

To do so it has several paths at option (in order of preference: the less intermediaries, the better):

1. If client X has a free LAN interface: Connect by cable and link to a pseudo (static for Odoo Pos Boxes) link-local IP *, **
2. If client X has a free WLAN interface: Connect to WLAN "pospox" and link to pseudo (static for Odoo Pos Boxes) link-local IP *. **
3. If the client X has neither a free WLAN nor a free LAN interface***: Probe the mDNS service discovery for `com.odoo.posbox.hash(MAC)` (sic!) - conigure that MAC in the backend ****
4. If the client X has neither a free WLAN nor a free LAN interface*** and 3 fails after x seconds: Connect to a Router and link to a configurable IP of the Raspberry LAN interface
5. If the client X has neither a free WLAN nor a free LAN interface*** and 4 fails after x seconds: Connect to a Router and link to a configurable IP of the Raspberry WLAN interface.

- Make 3 & 4 both configurable on the backend to always have a backup at hand.
- MAC should be optional on the backend, and if not set, it should discover service `com.odoo.posbox` which is bound by every raspberry in the reachable network (case: you only have one raspberry)


\* JustWorks (TM)
\** You don't waive broader network access (eg for ngrok-ing) with network sharing on Client X
\*** Detected in POS web client's reconciliation loop through probing the preconfigured pseudo link-local addresses at  `/hw_posbox/hello`
\**** JustWorks (TM) on networks where you don't have access to the Router (because it's provided and protected by ISP or the network admin is mean)


How does that sound to you?

**Comment by jorenvo** [comment]
Perhaps in the future we might be doing something so the user doesn't have to manually set the IP (using zeroconf or something similar). But this would not be to avoid DHCP server issues, just to make it easier to use. In fact, for that to work you still need DHCP.

I don't fully understand your steps, in 1) you want to plug an ethernet cable directly from a computer to the POSBox? I'm not sure what that would achieve.

Is the goal of all of this still to work around a faulty DHCP server?

**Comment by blaggacao** [comment]
>Is the goal of all of this still to work around a faulty DHCP server?

Exactly, when _CM_ has two interfaces one could still be available for a link local connection (can be LAN or WLAN) to directly connect to the _RP_ and operate independently of any DHCP or network topology. Result: JustWorks(TM).

>In fact, for that to work you still need DHCP.
Actually not. I think RFC3927 suggests the contrary:
https://tools.ietf.org/html/rfc3927
See also: https://tools.ietf.org/html/rfc4862

**Comment by Yenthe666** [comment]
Just my five cents but this totally feels like overengineering.

**Comment by jorenvo** [comment]
> > Is the goal of all of this still to work around a faulty DHCP server?

> Exactly,

Ok, in that case I don't think it's something we want to support, sorry. It's like buying an Android tablet and complaining wifi doesn't work when DHCP is down.

> I have in mind the LAN network cable which goes through the stairways up to the first floor is a possible failure point when a worker steps on it once a week.

When deploying a POSBox you have to have a reliable LAN network, and time should be spent improving the network rather than applying all kind of hacks to a single client to try and make it work. If your LAN network is unreliable you will run into issues with all devices (not just the POSBox) and the user experience will be bad regardless.

**Comment by blaggacao** [comment]
> When deploying a POSBox you have to have a reliable LAN network, and time should be spent improving the network rather than applying all kind of hacks to a single client to try and make it work. If your LAN network is unreliable you will run into issues with all devices (not just the POSBox) and the user experience will be bad regardless.

I would agree, but some third party would not. So don't shoot the messenger. :wink: 

> Ok, in that case I don't think it's something we want to support, sorry. It's like buying an Android tablet and complaining wifi doesn't work when DHCP is down.

Hey, that's part of your client base around the world. Shame at them! :wink: But it's actually a little different (if you factor in subjectively available knowledge): It's like buying an Android tablet and complaining wifi doesn't work.

> Just my five cents but this totally feels like overengineering.

It might feel so, but it's not. This is common sense in IOT engineering. You **have** to account for common failure scenarios... which is why I'm not really happy with:

> Ok, in that case I don't think it's something we want to support, sorry.

**Comment by blaggacao** [comment]
Btw, for anyone who stumbles over this: you can emulate step 2 of https://github.com/odoo/odoo/issues/25666#issuecomment-403993842 perfectly via the `10.11.12.1` hard coded address. It's not officially support, but comes pretty close to circumvent local network problems.

**Comment by osasidiaru** [comment]
I've never heard of DHCP servers behaving in this way, nor do I understand the benefit of forcing a client to release a valid lease. Why not just allow it to use it? What happens if a client somehow loses this lease because they clear some data? Then the client can no longer connect until that lease expires on the DHCP server? Could you point to some flag in isc-dhcp-server that enables this crazy policy?

**Comment by blaggacao** [comment]
@osasidiaru Had I genuine access to the target system, this would probably be no issue no more. I'm not sure if this is considered a valid course of action, but I'm trying - from the outside - through educated guesswork to solve a real client's issue. And after acknowledging that it probably could not be solved properly, I tried to find a reconciliation loop which would help mitigate those problems (which include human aspects, as stress, uncertainty, non-observability) from a holistic point of view to reach the ultimate goal of proper BCM for such a vital component as a POS.
